src/shapiro_wilk.jl

```diff
@@ -0,0 +1,260 @@
+Base.IndexStyle(::Type{ShapiroWilkCoefs}) = IndexLinear()
+
+Base.@propagate_inbounds function Base.getindex(SWc::ShapiroWilkCoefs, i::Integer)
+    @boundscheck checkbounds(SWc, i)
+    @inbounds if checkbounds(Bool, SWc.A, i)
+        return SWc.A[i]
+    elseif isodd(length(SWc)) && i == lastindex(SWc.A) + 1
+        return zero(eltype(SWc))
+    else
+        return -SWc.A[SWc.N + 1 - i]
+    end
+end
+
+function ShapiroWilkCoefs(N::Integer)
+    if N < 3
+        throw(ArgumentError("N must be greater than or equal to 3, got $N"))
+    elseif N == 3 # exact
+        return ShapiroWilkCoefs(N, [sqrt(2.0) / 2.0])
+    else
+        # Weisberg&Bingham 1975 statistic; store only positive half of m:
+        # it is (anti-)symmetric; hence '2' factor below
+        m = [-quantile(Normal(), (i - 3 / 8) / (N + 1 / 4)) for i in 1:div(N, 2)]
+        mᵀm = 2sum(abs2, m)
+        x = 1 / sqrt(N)
+        a₁ = m[1] / sqrt(mᵀm) + __RS92_C1(x) # aₙ = cₙ + (...)
+        if N ≤ 5
+            ϕ = (mᵀm - 2m[1]^2) / (1 - 2a₁^2)
+            m .= m ./ sqrt(ϕ) # A, but reusing m to save allocs
+            m[1] = a₁
+        else
+            a₂ = m[2] / sqrt(mᵀm) + __RS92_C2(x) # aₙ₋₁ = cₙ₋₁ + (...)
+            ϕ = (mᵀm - 2m[1]^2 - 2m[2]^2) / (1 - 2a₁^2 - 2a₂^2)
+            m .= m ./ sqrt(ϕ) # A, but reusing m to save allocs
+            m[1], m[2] = a₁, a₂
+        end
+        return ShapiroWilkCoefs(N, m)
+    end
+end
+
+function unsafe_swstat(X::AbstractVector{<:Real}, A::ShapiroWilkCoefs)
+    AX = dot(view(A, 1:length(X)), X)
+    m = mean(X)
+    S² = sum(x -> abs2(x - m), X)
+    W = AX^2 / S²
+    return min(W, one(W)) # to guard against numeric errors
+end
+
+struct ShapiroWilkTest <: HypothesisTest
+    coefs::ShapiroWilkCoefs # expectation of order statistics for Shapiro-Wilk test
+    W::Float64              # test statistic
+    censored::Int           # only the smallest N₁ samples were used
+end
+
+testname(::ShapiroWilkTest) = "Shapiro-Wilk normality test"
+function population_param_of_interest(t::ShapiroWilkTest)
+    return ("Squared correlation of data and expected order statistics of N(0,1) (W)",
+            1.0, t.W)
+end
+default_tail(::ShapiroWilkTest) = :left
+censored_ratio(t::ShapiroWilkTest) = t.censored / length(t.coefs)
+
+function show_params(io::IO, t::ShapiroWilkTest, indent)
+    l = 24
+    println(io, indent, rpad("number of observations:", l), length(t.coefs))
+    println(io, indent, rpad("censored ratio:", l), censored_ratio(t))
+    return println(io, indent, rpad("W-statistic:", l), t.W)
+end
+
+function StatsAPI.pvalue(t::ShapiroWilkTest)
+    n = length(t.coefs)
+    W = t.W
+
+    if iszero(censored_ratio(t))
+        if n == 3 # exact by Shapiro&Wilk 1965
+            # equivalent to 6/π * (asin(sqrt(W)) - asin(sqrt(3/4)))
+            return 1 - 6acos(sqrt(W)) / π
+        elseif n ≤ 11 # Royston 1992
+            γ = __RS92_G(n)
+            if γ ≤ log1p(-W)
+                return zero(W)
```

Comment:
@nalimilan This line is uncovered by tests; I tried to find an example two weeks ago as well. I remember that I encountered a floating point issue here, but maybe `log1p` doesn't suffer from it?

I fuzzed `pvalue` for 20 minutes with `rand` and `randn` samples but did not find a counterexample.
If you consider this sufficient we could remove this line, although in Royston 1992 there is no proof that `__RS92_G` is an upper bound for `log1p(-W)`

Reply:
Do you mean you can't find values that allow covering this branch? That would be annoying.

If you're just unsure whether this branch is useful, I'd prefer erring on the safe side and keeping it if it's in Royston's algorithm. I'm not sure random number are a good way to check correctness, as problems can appear only in very rare case (e.g. subnormal numbers).

Reply:
So I've run some BBOptim on this problem and here are the results:
```julia
using BlackBoxOptim

function wstat(x)
    n = length(x)
    W = ShapiroWilkTest(x).W
    γ = HypothesisTests.__RS92_G(n)
    return γ - log1p(-W)
end

map(4:11) do d
    range = (-1.0,1.0)
    res = bboptimize(wstat, NumDimensions=d, SearchRange = range)
    best_fitness(res)
end

8-element Vector{Float64}:
 0.5566477668605001
 0.8253711855266714
 1.1663663366212265
 1.5432534398612425
 1.940969686090396
 2.352036190063946
 2.7722658295384175
 3.1991066183178085

```
I also had a look at the points returned by the optimizer and they seem to follow "common-sense" pattern for non-normality: e.g. `[-1.0, 1.0, ..., 1.0]`
(note that `W` is scale- and shift-invariant so search-range shouldn't matter).

While this is not a proof, it seems that this if-branch is dead and can be removed.
The branch is not mentioned in the articles of Royston, however it is present in the original Fortran (`swilk.f`).

Reply:
@nalimilan should we remove this branch (I'd argue for this), or keep it?
In the latter case this seems ready to be merged ;)

Reply:
Sorry, I missed your reply. Yeah let's drop it if you're confident it's never used.